core/views.py

```diff
@@ -66,6 +69,14 @@ class PostCreateView(LoginRequiredMixin, CreateView):
     template_name = "core/post-create.html"
     login_url = "login"
 
+    def dispatch(self: "PostCreateView", request: HttpRequest, *args: list, **kwargs: dict[str, Any]) -> HttpResponse:
+        if not request.user.is_authenticated:
+            return redirect(reverse("login"))
+        if not request.user.create_post:
+            messages.error(request, "You do not have permission to create posts.")
+            return redirect(reverse("post-list"))
+        return super().dispatch(request, *args, **kwargs)
+
     def get_form_kwargs(self: "PostCreateView") -> dict[str, any]:
```

Comment:
Let's not annotate `self` here and in the entire PR (`self` and `cls` types are inferred by mypy to proper forms of `typing.Self`). We also don't need to use the quotes for deferring annotations—see [PEP 563](https://peps.python.org/pep-0563/).
I don't know how it looks in the rest of the codebase, but I can make sure it coherently sticks to PEP 563 and no self annotations.